src/org/intellij/erlang/compilation/ErlangPrepareDependenciesCompileTask.java

```diff
@@ -127,162 +128,147 @@ private static ErlangModuleBuildOrderDescriptor getModuleBuildOrderInner(Module
+    builder.buildDFST();
+    if (!builder.isAcyclic()) {
+      throw new CyclicDependencyFoundException();
+    }
+    return ContainerUtil.map(builder.getSortedNodes(), new Function<String, ErlangFileDescriptor>() {
+      @Override
+      public ErlangFileDescriptor fun(String filePath) {
+        Collection<String> dependencies = semiGraph.getDependencies(filePath);
+        assert dependencies != null;
+        return new ErlangFileDescriptor(filePath, dependencies);
+      }
+    });
   }
 
-  private static class ErlangModulesSorter {
-    private final Collection<ErlangFile> myModules;
-    private final List<String> myGlobalParseTransforms;
-
-    private ErlangModulesSorter(Collection<ErlangFile> modules, List<String> globalParseTransforms) {
-      myModules = modules;
-      myGlobalParseTransforms = globalParseTransforms;
+  private static class ErlangFilesDependencyGraph implements GraphGenerator.SemiGraph<String> {
+    private final HashMap<String, String> myNamesToPathsMap = ContainerUtil.newHashMap();
+    private final HashMap<String, Collection<String>> myPathsToDependenciesMap = ContainerUtil.newHashMap();
+    private final PsiManager myPsiManager;
+
+    public ErlangFilesDependencyGraph(@NotNull Collection<VirtualFile> erlangModules,
+                                      @NotNull Collection<VirtualFile> erlangHeaders,
+                                      @NotNull List<String> globalParseTransforms,
+                                      @NotNull PsiManager psiManager) {
+      myPsiManager = psiManager;
+      buildNamesMap(erlangModules);
+      buildForHeaders(erlangHeaders);
+      buildForModules(erlangModules, globalParseTransforms);
     }
 
-    public static List<ErlangFileDescriptor> sort(Collection<ErlangFile> erlangModules,
-                                                  List<String> globalParseTransforms) throws CyclicDependencyFoundException {
-      ErlangModulesSorter sorter = new ErlangModulesSorter(erlangModules, globalParseTransforms);
-      return sorter.getSortedModules();
+    private void buildForModules(@NotNull Collection<VirtualFile> erlangModules,
+                                 @NotNull List<String> globalParseTransforms) {
+      for (VirtualFile erlangModule : erlangModules) {
+        Set<String> dependencies = ContainerUtil.newHashSet();
+        ErlangFile erlangFile = getErlangFile(erlangModule);
+        assert erlangFile != null : "We excepted correct PsiFile for " + erlangModule.getPath();
+        addDeclaredDependencies(erlangFile, dependencies);
+        dependencies.addAll(getPathsFromNames(globalParseTransforms));
+        myPathsToDependenciesMap.put(erlangModule.getName(), dependencies);
+      }
     }
 
-    private GraphGenerator<Node> createModulesGraph() {
-      return GraphGenerator.create(new ErlangModulesDependencyGraph(myModules, myGlobalParseTransforms));
+    @Nullable
+    private ErlangFile getErlangFile(@NotNull VirtualFile virtualFile) {
```

Comment:
Why is it Nullable? You only call it to obtain a not-null value. Move your assertion here, make the method not null.

Also, make use of ObjectUtils.tryCast - it will improve readability.